src/pybind/mgr/dashboard/controllers/rbd.py

```diff
@@ -1,136 +1,422 @@
+                                 "journaling"])
+
+    # set of image features that can be disabled on existing images
+    ALLOW_DISABLE_FEATURES = set(["exclusive-lock", "object-map", "fast-diff",
+                                  "deep-flatten", "journaling"])
+
+    @classmethod
+    def _rbd_disk_usage(cls, image, snaps):
+        class DUCallback(object):
+            def __init__(self):
+                self.used_size = 0
+
+            def __call__(self, offset, length, exists):
+                if exists:
+                    self.used_size += length
+
+        snap_map = {}
+        prev_snap = None
+        total_used_size = 0
+        for _, size, name in snaps:
+            image.set_snap(name)
+            du_callb = DUCallback()
+            image.diff_iterate(0, size, prev_snap, du_callb, whole_object=True)
+            snap_map[name] = du_callb.used_size
+            total_used_size += du_callb.used_size
+            prev_snap = name
+
+        return total_used_size, snap_map
+
+    def _rbd_image(self, ioctx, pool_name, image_name):
+        img = rbd.Image(ioctx, image_name)
+        stat = img.stat()
+        stat['name'] = image_name
+        stat['id'] = img.id()
+        stat['pool_name'] = pool_name
+        features = img.features()
+        stat['features'] = features
+        stat['features_name'] = _format_bitmask(features)
+
+        # the following keys are deprecated
+        del stat['parent_pool']
+        del stat['parent_name']
+
+        stat['timestamp'] = "{}Z".format(img.create_timestamp().isoformat())
+
+        stat['stripe_count'] = img.stripe_count()
+        stat['stripe_unit'] = img.stripe_unit()
+
+        data_pool_name = CephService.get_pool_name_from_id(img.data_pool_id())
+        if data_pool_name == pool_name:
+            data_pool_name = None
+        stat['data_pool'] = data_pool_name
+
+        try:
+            parent_info = img.parent_info()
+            stat['parent'] = {
+                'pool_name': parent_info[0],
+                'image_name': parent_info[1],
+                'snap_name': parent_info[2]
+            }
+        except rbd.ImageNotFound:
+            # no parent image
+            stat['parent'] = None
+
+        # snapshots
+        stat['snapshots'] = []
+        for snap in img.list_snaps():
+            snap['timestamp'] = "{}Z".format(img.get_snap_timestamp(snap['id']).isoformat())
+            snap['is_protected'] = img.is_protected_snap(snap['name'])
+            snap['used_bytes'] = None
+            snap['children'] = []
+            img.set_snap(snap['name'])
+            for child_pool_name, child_image_name in img.list_children():
+                snap['children'].append({
+                    'pool_name': child_pool_name,
+                    'image_name': child_image_name
+                })
+            stat['snapshots'].append(snap)
+
+        # disk usage
+        if 'fast-diff' in stat['features_name']:
```

Comment:
... should also verify that the fast-diff isn't invalid via ```flags()``` 

[1] https://github.com/ceph/ceph/blob/master/src/pybind/rbd/rbd.pyx#L78